h2o-core/src/main/java/hex/grid/FilteredHyperSpaceIterator.java

```diff
@@ -0,0 +1,81 @@
+package hex.grid;
+
+import hex.Model;
+import hex.grid.filter.PermutationFilter;
+
+public class FilteredHyperSpaceIterator<MP extends Model.Parameters> implements HyperSpaceWalker.HyperSpaceIterator<MP> {
+  private final HyperSpaceWalker.HyperSpaceIterator<MP> _hyperSpaceIterator;
+
+  private final PermutationFilter<MP> _permutationFilter;
+
+  /**
+   * Total number of visited permutations, including those filtered out by {@code _permutationFilter}.
+   */
+  private long _numberOfVisitedPermutations = 0;
+
+  /**
+   * Keeps number of returned to the user permutations as not all visited permutations are considered to be worthy
+   * for evaluation due to a {@code _permutationFilter}.
+   */
+  private long _numberOfUsedPermutations = 0;
+  private final long _maxHyperSpaceSize;
+
+  public FilteredHyperSpaceIterator(HyperSpaceWalker.HyperSpaceIterator<MP> hyperSpaceIterator, PermutationFilter<MP> permutationFilter, long maxHyperSpaceSize) {
+    _hyperSpaceIterator = hyperSpaceIterator;
+    _permutationFilter = permutationFilter;
+    _maxHyperSpaceSize = maxHyperSpaceSize;
+  }
+
+  @Override
+  public MP nextModelParameters(Model previousModel) {
+    MP permutation = _hyperSpaceIterator.nextModelParameters(previousModel);
+
+    while (permutation != null && !_permutationFilter.test(permutation)) {
+      _numberOfVisitedPermutations++;
+      permutation = hasNext(previousModel) ? _hyperSpaceIterator.nextModelParameters(previousModel) : null;
+    }
+    _numberOfVisitedPermutations++;
+    _numberOfUsedPermutations++;
+    return permutation;
+  }
+
+  @Override
+  public boolean hasNext(Model previousModel) {
+    return _numberOfVisitedPermutations < _maxHyperSpaceSize && ( max_models() == 0 || _numberOfUsedPermutations < max_models());
```

Comment:
I don't think the iterator cannot make an assumption on what parameters will be actually used to make models. 

`hasNext` is currently used like this:
```
      while (it.hasNext(model) && (it.max_models() > 0 && count++ < it.max_models())) { //only walk the first max_models models, if specified
```

your `hasNext` duplicates that, imho it shouldn't be here

Reply:
@michalkurka
Not completely understand your point. Maybe you used double negation `don't` and `cannot` by mistake? Anyway, could you please clarify?

 it does not duplicate, as it uses `used` counter instead of `visited`. I think it is correct to return expected max_models number of parameters even if some were filtered out along the search. I was trying to partially reuse part of the condition, but it is probably not possible.

Reply:
> Not completely understand your point. Maybe you used double negation don't and cannot by mistake? Anyway, could you please clarify?

yes

max_models is used by the caller to limit the models, the iterator shouldn't use max models because of filtering that can be done on other levels - your design allows for chaining of the iterators and filtering

=> you shouldn't stop early because you reached max_model - there can be something else filtering it

Reply:
@michalkurka but I'm not stopping earlier. I'm using max_models from underneath walker. And as a result I will stop later. Sorry, still don't see an issue

Reply:
but I'm not giving up. Trying to get your point.

Reply:
user asks for up to 20 models, the design allows for multi-level filtering - I can nest your FilteredHyperSpaceIterator into another filter, if your stops after 20, and my filter filters out some more you won't get 20 models

my point was that in the current code base the decision to stop is done on the top level which makes your check redundant and perhaps even incorrect if I want to nest the iterator into another filtering iterator

Reply:
are you referring to the fact that I don't have `_hyperSpaceIterator.hasNext(previousModel);` logic to delegate? if yes... I was trying to do it this way but logic of iteration and early stopping is intertwined that makes it impossible to use from FilterIterator. It should be separated then in original iterator. I tried but then decided that overriding completely could be fine as well. But if you want to add another filter iterator in between original iterator and this one then yes... it will not work. But I don't think we need to consider that scenario

Reply:
max_models is only considered here in the current code base: https://github.com/h2oai/h2o-3/blob/master/h2o-core/src/main/java/hex/grid/GridSearch.java#L122

`hasNext` should just say if there are any other parameters, stopping on max models will be handled by gridsearch

Reply:
@michalkurka 

> max_models is only considered here in the current code base:

What about this usage https://github.com/h2oai/h2o-3/blob/39ace2cf92ea362f52cf13ad3d82cfab762e9c6c/h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java#L538 ?